src/inventory.cpp

```diff
@@ -782,11 +782,51 @@ ItemStack InventoryList::peekItem(u32 i, u32 peekcount) const
 	return m_items[i].peekItem(peekcount);
 }
 
-void InventoryList::moveItem(u32 i, InventoryList *dest, u32 dest_i, u32 count)
+void InventoryList::moveItemSomewhere(u32 i, InventoryList *dest, u32 count)
 {
-	if(this == dest && i == dest_i)
+	// Take item from source list
+	ItemStack item1;
+	if (count == 0)
+		item1 = changeItem(i, ItemStack());
+	else
+		item1 = takeItem(i, count);
+
+	if (item1.empty())
 		return;
 
+	// Try to add the item to destination list
+	u32 oldcount = item1.count;
+	u32 dest_size = dest->getSize();
+	// First try all the non-empty slots
+	for (u32 dest_i = 0; dest_i < dest_size; dest_i++) {
+		if (!m_items[dest_i].empty()) {
+			item1 = dest->addItem(dest_i, item1);
+			if (item1.empty()) break;
+		}
+	}
+
+	// Then try all the empty ones
+	for (u32 dest_i = 0; dest_i < dest_size; dest_i++) {
+		if (m_items[dest_i].empty()) {
+			item1 = dest->addItem(dest_i, item1);
+			if (item1.empty()) break;
```

Comment:
If we break if item1 is empty and next we look if item1 is not empty, you can replace break with return here and remove the next condition